# TL - 230519
library(ggplot2)

setwd("/media/thibaultleroy/Seagate Expansion Drive/projet_BNP-BHW/BayPass_100kSNPs")

###### distribution of XtX observed values ####################################
realdata=read.table("BayPass_merged_results_121017_Toc_Pluvio.txt.pseudoK.withheader",h=T)
# this file is a summary of the SNP-specific output files generated by baypass
#scaff   pos     ref     xtx     1MeanToc        BFMeanToc       eBFMeanToc      2MeanPluvio     BFMeanPluvio    eBFMeanPluvio   chr     pos
#Sc0000000       603     C       21.35075058     1       -9.94543070     0.46568302      2       -11.98530974    0.20408554      Chr1    20675814
#Sc0000000       8460    T       17.58852127     1       -4.41161702     0.15798716      2       -10.78520360    0.20779499      Chr1    20683671
#...

ggplot(data=realdata,aes(x=xtx)) + geom_density(size=1.4) + xlab("Observed XtX values")+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))+ theme(legend.position = "none")+
  theme(legend.title=element_text(size=12, face="bold"), legend.title.align=0.5, legend.text=element_text(size=10, face="italic"), axis.line = element_line(colour = 'black', size = 1.25), axis.ticks = element_line(colour = 'black', size = 1.25), 
        axis.text.x = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.5,face="plain"),
        axis.text.y = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.5,face="plain"),  
        axis.title.x = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.2,face="italic"),
        axis.title.y = element_text(colour="black",size=12,angle=90,hjust=.5,vjust=.5,face="italic"))



### distribution of XtX values based on PODS (pseudo-observed datasets) #############

# Perform neutral simulations assuming the variance-covariance matrix (BayPass Manual & Gautier et al. 2015)
source("~/Software/baypass_2.1/utils/baypass_utils.R") # import Baypass-associated R functions

# read the omega matrix (variance-covariance matrix)
om.bta=as.matrix(read.table("BNP-BHW_18pops_v2.3_freqSNP_mincount10_filterd_biallelic_sites_counts.1.baMeanAnTocMeanPluvio_mat_omega.out"))
pi.beta.coef=read.table("BNP-BHW_18pops_v2.3_freqSNP_mincount10_filterd_biallelic_sites_counts.1.baMeanAnTocMeanPluvio_summary_beta_params.out",h=T)$Mean
geno.data<-geno2YN("BNP-BHW_18pops_v2.3_freqSNP_mincount10_filterd_biallelic_sites_counts.1.genobaypass")
simu.bta<-simulate.baypass(omega.mat=om.bta,nsnp=100000,sample.size=geno.data$NN,
                           beta.pi=pi.beta.coef,pi.maf=0.02,coverage=150,suffix="BNP-BHW_18pops_v2.3_freqSNP_mincount10_filterd_biallelic_sites_counts.PODS.1")


# XtX values based on simulations.
pods=read.table("BNP-BHW_18pops_v2.3_freqSNP_mincount10_filterd_biallelic_sites_counts.PODS.1_summary_pi_xtx.out",h=T)
head(pods)
ggplot(data=pods,aes(x=M_XtX)) + geom_density(size=1.4) + xlab("Simulated XtX values")+
    scale_color_manual(values = c(rep("grey",153)))+
    theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))+ theme(legend.position = "none")+
    theme(legend.title=element_text(size=12, face="bold"), legend.title.align=0.5, legend.text=element_text(size=10, face="italic"), axis.line = element_line(colour = 'black', size = 1.25), axis.ticks = element_line(colour = 'black', size = 1.25), 
    axis.text.x = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.5,face="plain"),
    axis.text.y = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.5,face="plain"),  
    axis.title.x = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.2,face="italic"),
    axis.title.y = element_text(colour="black",size=12,angle=90,hjust=.5,vjust=.5,face="italic"))
# compute threshold based on the simulations
simulatedxtx=quantile(pods$M_XtX,probs=c(0.99,0.999,0.9999,0.99999))
#     99%    99.9%   99.99%  99.999% 
#20.22704 22.62722 25.83772 27.89115
# How to interpret: 99% of the variant simulated exhibit XtX values lower than 20.227


#### Expected vs. observed XtX distributions #############
setwd("/home/thibaultleroy/Collab/Sessile_Reanalyse/")
#tiff("XtX distributions.tiff", width = 10, height = 6, units = 'in', res = 600)
png("XtX distributions.png", width = 10, height = 6, units = 'in', res = 600)
#jpeg("XtX distributions.jpeg", width = 10, height = 6, units = 'in', res = 600)

ggplot(data=pods,aes(x=M_XtX)) + geom_density(size=1.4) + geom_density(data=realdata,aes(x=xtx),col="red") + xlab("XtX values")+ xlim(0,60)+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"))+ theme(legend.position = "none")+
  theme(legend.title=element_text(size=12, face="bold"), legend.title.align=0.5, legend.text=element_text(size=10, face="italic"), axis.line = element_line(colour = 'black', size = 1.25), axis.ticks = element_line(colour = 'black', size = 1.25), 
        axis.text.x = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.5,face="plain"),
        axis.text.y = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.5,face="plain"),  
        axis.title.x = element_text(colour="black",size=12,angle=0,hjust=.5,vjust=.2,face="italic"),
        axis.title.y = element_text(colour="black",size=12,angle=90,hjust=.5,vjust=.5,face="italic"))+
  # add annotations for the quantiles
  geom_segment(x=20.227, y=0, xend=20.227, yend=0.3, size=0.5, linetype="dotted",color="black")+
  annotate("text", x=20.2, y=0.35, label= "99%") + 
  geom_segment(x=27.89115, y=0, xend=27.89115, yend=0.2, size=0.3, linetype="dotted",color="black")+
  annotate("text", x=27.8, y=0.25, label= "99.999%")  

dev.off()


#### Generate a Manhattan plot of XtX values #########################
setwd("/media/thibaultleroy/Seagate Expansion Drive/projet_BNP-BHW/BayPass_100kSNPs")
realdata2=read.table("BayPass_merged_results_121017_Toc_Pluvio.txt.pseudoK.withheader",h=T)
head(realdata2)
bidon=as.data.frame(head(realdata2))
#png("PCAdapt_scan_chr1-6.png", width = 6, height = 8, units = 'in', res = 600)
par(mfrow = c(3, 2))
for (i in 1:6){ # first 6 chromosomes
  datacurrentchr=subset(realdata2,realdata2$chr==paste("Chr",i,sep=""))
  plot(datacurrentchr$xtx~datacurrentchr$pos.1,main=paste("Chromosome ",i),pch=20,xlab=paste ("position - chr", i), ylab="XtX",col=ifelse(datacurrentchr$xtx>27.89115,"black",ifelse(datacurrentchr$xtx>20.22704,"grey40","grey")),cex=ifelse(datacurrentchr$xtx>27.89115,0.02,ifelse(datacurrentchr$xtx>20.22704,0.002,0.0001)))+
  abline(20.22704,0,col="orange",lwd=2,lty=2)
  abline(27.89115,0,col="red",lwd=2,lty=2)
}
#dev.off()

for (i in 7:12){ # last 6 chromosomes
  datacurrentchr=subset(realdata2,realdata2$chr==paste("Chr",i,sep=""))
  plot(datacurrentchr$xtx~datacurrentchr$pos.1,main=paste("Chromosome ",i),pch=20,xlab=paste ("position - chr", i), ylab="XtX",col=ifelse(datacurrentchr$xtx>27.89115,"black",ifelse(datacurrentchr$xtx>20.22704,"grey40","grey")),cex=ifelse(datacurrentchr$xtx>27.89115,0.02,ifelse(datacurrentchr$xtx>20.22704,0.002,0.0001)))+
    abline(20.22704,0,col="orange",lwd=2,lty=2)
  abline(27.89115,0,col="red",lwd=2,lty=2)
}
